### PR TITLE
Spark Executor Tasks Instrumentation

### PR DESCRIPTION
**Pull request by paul-laffon-dd**
# What Does This Do

Add a new instrumentation `spark-executor` to trace tasks running on spark executors

The existing spark instrumentation is running on the spark driver, capturing spark internals (spark.sql, spark.job, spark.stage...), while this new instrumentation will trace task being run on spark executors.

# Motivation

Generate traces for `spark.task`, with underlying spans (like s3 or database calls) being linked. 

# Additional Notes

- each `spark.task` is a distinct trace and not on the existing `spark.application` / `spark.job` traces:
  - a spark job/application can have millions on tasks, making it unpractical to have everything in one trace
  - sampling should be applied to `spark.task`, while all `spark.application` and `spark.job` traces are kept  
  - Span link will be used to link the `spark.task` to other spark traces

- task errors are not captured for now
  - spark [catch all throwable](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L801-L841) in task execution, making it complex to retrieve it from the instrumentation

### CODE REVIEW

**Review by raphaelgavache — APPROVED** [review]
lgtm, looks great!

**Review by bm1549 — COMMENTED** [review]
Can you also update CODEOWNERS so that [data-jobs-monitoring](https://github.com/orgs/DataDog/teams/data-jobs-monitoring) owns the instrumentation files here?

**Comment by paul-laffon-dd** [comment]
@bm1549 In order to add the team [@DataDog/data-jobs-monitoring](https://github.com/orgs/DataDog/teams/data-jobs-monitoring) as CODEOWNERS, it need to have access to the repository

image

**Comment by bm1549** [comment]
@paul-laffon-dd it seems like that team is not a subteam of https://github.com/orgs/DataDog/teams/dev which is causing the issue

@tiffon as maintainer of the team, can you update the parent team to be dev?
https://github.com/orgs/DataDog/teams/data-jobs-monitoring/edit

**Comment by paul-laffon-dd** [comment]
Thanks @bm1549, we updated the team settings and the CODEOWNERS file is now valid

**Review by dougqh — CHANGES_REQUESTED** [review]
Scope.close and span.finish should probably be in finally blocks; otherwise, you risk traces not being published in a timely fashion.

File: dd-java-agent/instrumentation/spark-executor/src/main/java/datadog/trace/instrumentation/spark/SparkExecutorInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        @Advice.Enter final AgentScope scope, @Advice.This final Executor.TaskRunner taskRunner) {
+      if (scope == null) {
+        return;
+      }
+
+      AgentSpan span = scope.span();
+      DECORATE.onTaskEnd(span, taskRunner);
+      DECORATE.beforeFinish(scope);
+
+      scope.close();
```
[the remark this answers is not in the document]
Comment (paul-laffon-dd) [reply dd-java-agent/instrumentation/spark-executor/src/main/java/datadog/trace/instrumentation/spark/SparkExecutorInstrumentation.java]: Move those in finally block

File: dd-java-agent/instrumentation/spark-executor/src/main/java/datadog/trace/instrumentation/spark/SparkExecutorInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return;
+      }
+
+      final AgentSpan span = scope.span();
+
+      try {
+        DECORATE.onTaskEnd(span, taskRunner);
+        DECORATE.beforeFinish(scope);
+      } finally {
+        scope.close();
```
Comment (dougqh) [inline dd-java-agent/instrumentation/spark-executor/src/main/java/datadog/trace/instrumentation/spark/SparkExecutorInstrumentation.java:77]: I think this is probably fine, but typically each clean-up needs to be in its own finally block.
In this case - if scope.close raises an exception then you won't perform span.finish.

**Review by dougqh — APPROVED** [review]

**Pull request merged**